### ISSUE TITLE
Template.subscribe with DDP.connect

### ISSUE THREAD

**Issue by jamgold** [meteor/meteor#3959]
The new feature in 1.0.4 to use subscription inside templates does not work when providing a remote connection created with DDP.connect.

According to `Blaze.View.prototype.subscribe` it should be possible to pass in options defining a connection, but no matter how I do it, I can't get it to work.

Worst case scenario the browser crashes with Maximum call stack size exceeded, and best case it doesn't subscribe at all or throws an error 
```
Meteor.makeErrorType.errorClass {
 error: 404,
 reason: "Subscription not found",
 details: undefined,
 message: "Subscription not found [404]",
 errorType: "Meteor.Error"…
}
```

**Comment by glasser** [comment]
Please provide a full reproduction as described in: https://github.com/meteor/meteor/wiki/Contributing-to-Meteor#reporting-a-bug-in-meteor  Without a way to replicate your problem, we can't debug and fix it.

**Comment by jamgold** [comment]
Understood, here is the reproduction https://github.com/jamgold/meteor-issue-3959

**Comment by stubailo** [comment]
Please change this line: https://github.com/jamgold/meteor-issue-3959/blob/1049a2c31e34bec4096faeb1ad11e9bc5fc5d608/tsub.js#L20

To be `this.view.subscribe`, and pass the arguments in an array, like I mentioned in the forum post: https://forums.meteor.com/t/meteor-1-0-4-template-subscribe-and-ddp-connect/1532/6?u=sashko

**Comment by jamgold** [comment]
When you do that with the reproduction I submitted, you are telling me it works? Because it sure does __not__ work when I run it on my server.

**Comment by stubailo** [comment]
https://github.com/jamgold/meteor-issue-3959/pull/1

Here's a PR.

I think I wasn't clear that the "arguments" include the options to `Meteor.subscribe` itself. In fact, this API should be improved to merge the options of `this.view.subscribe` and `Meteor.subscribe` probably.

**Comment by jamgold** [comment]
Can you make that work __without__ changing the DDP.connect url to / ?

**Comment by stubailo** [comment]
That's an issue unrelated to `template.subscribe` - in this case we should figure out why SockJS can't make a websocket connection to your server.

**Comment by jamgold** [comment]
but it can, if you use Meteor.remoteConnection.subscribe ...

**Comment by stubailo** [comment]
From my home computer, I get this error:

```
WebSocket connection to 'ws://mongodb-image.devel.buzzledom.com/sockjs/003/trm_zdim/websocket' failed: One or more reserved bits are on: reserved1 = 0, reserved2 = 1, reserved3 = 1`
```

**Comment by stubailo** [comment]
Is that server a Meteor server?

**Comment by jamgold** [comment]
Behind phusion passenger, yes

**Comment by jamgold** [comment]
Try it with http://mongdb-image.devel.buzzledom.com:3000

**Comment by stubailo** [comment]
OK, I see - it looks like the regular method works but through `this.view` doesn't. Sorry about that.

**Comment by jamgold** [comment]
No problem, I thought I am doing something wrong on my end ...

**Comment by jamgold** [comment]
The problem is within `Blaze.View.prototype.subscribe` in packages/blaze/view.js.
```
Blaze.View.prototype.subscribe = function (args, options) {
  var self = this;
  options = {} || options;
// ^^^^^^^ this line resets the content of options 
  self._errorIfShouldntCallSubscribe();

  var subHandle;
  if (options.connection) {
    subHandle = options.connection.subscribe.apply(options.connection, args);
  } else {
    subHandle = Meteor.subscribe.apply(Meteor, args);
  }

  self.onViewDestroyed(function () {
    subHandle.stop();
  });

  return subHandle;
};
```
by the time the code reaches the if clause, options is empty. Tested in Safari, Chrome and Firefox

**Comment by jamgold** [comment]
changing the line to 
```
  options = options == undefined ? {} : options;
```
fixes it. After that everything works as expected

**Comment by jamgold** [comment]
incidentally this is the **only** occurrence of `var = {} || var` in all of Meteor

**Comment by glasser** [comment]
Heh, good catch.  (I guess this is not tested!) Want to send a PR to fix it?

**Comment by glasser** [comment]
(though we usually just write `options = options || {}`)

**Comment by rhyslbw** [comment]
Nice work @jamgold

**Comment by jamgold** [comment]
Also, `Blaze.TemplateInstance.prototype.subscribe` can be fixed so the code can be used as advertised in the docs like so
```
  // args.push(callbacks);

  subHandle = self.view.subscribe.call(self.view, args, callbacks);
```
with that change one can do the following
```
Template.x.created = function() {
 this.subscribe('collection', {
  connection: DDP.connect(url),
  onReady: function() {console.log('ready')},
  onError: function(err) {console.log(err)}
 });
}

**Comment by jamgold** [comment]
Let me read up on what I have to do for the PR (sign away a kidney what)

**Comment by glasser** [comment]
I think it was intentional to not allow you to specify `connection` to TemplateInstance.subscribe (and only in view.subscribe.  Although @stubailo I didn't realize when reviewing this that it would mean that subscriptionsReady would ignore those subscriptions!

**Comment by jamgold** [comment]
One more thing: according to the docs it says you can pass an object to Template.subscribe containing onStop and onReady in lieu of a function

> Optional. May include onStop and onReady callbacks. If a function is passed instead of an object, it is interpreted as an onReady callback.

the problem is, that Template.subscribe ignores the onReady

```
Temaplate.x.created = function(){
 this.subscribe('collection', {
  onError: function(err) {console.log(err);}
  onReady: function() { console.log('I will never be called')}
 });
}
```

**Comment by stubailo** [comment]
Interesting. I thought I took extra care to call it, maybe that was just onError. Sounds like there are a few bugs to be ironed out here!

**Comment by glasser** [comment]
@jamgold Are you sure?  onReady is interpreted by `Meteor.subscribe` (or `options.connection.subscribe`); it should be in the last element of `args` in View.subscribe.

**Comment by jamgold** [comment]
@glasser, Meteor.subcribe works, but Template.subscribe does not call onReady

**Comment by glasser** [comment]
@jamgold Works for me.  Can you show this as a [cloneable reproduction](https://github.com/meteor/meteor/wiki/Contributing-to-Meteor#reporting-a-bug-in-meteor)?

**Comment by jamgold** [comment]
@glasser, can you show me how you get it to work in the context of Template.subscribe with a remote DDP connection?

**Comment by glasser** [comment]
To be clear, my test was not with an alternate connection, which template.subscribe is not intended to support.

**Comment by jamgold** [comment]
Even so, it does not work
```
Collection = new Mongo.collection('collection');
if(Meteor.isClient)
{
 Template.hello.created = function() {
  this.subscribe('collection',{
   onReady: function(){console.log('collection ready');
  });
 }
}
if(Meteor.isServer)
{
 Meteor.publish('collection', function(){
  return Collection.find();
 });
}

**Comment by jamgold** [comment]
OK, my changes to 'Blaze.TemplateInstance.prototype.subscribe' break this ... :(

**Comment by jamgold** [comment]
@glasser, I fixed it. `Blaze.TemplateInstance.prototype.subscribe` will need to check for connection as well and separate it out. I will sign the contribution agreement later and fire off a pull request after work, if you still think this is worthwhile.

**Comment by stubailo** [comment]
Yep, I think you should make a PR - definitely a legitimate bug that we should fix.

**Comment by RobertLowe** [comment]
Nice catch!

**Comment by stubailo** [comment]
I think this was fixed in #4099
